src/Exchange.ts

```diff
@@ -286,75 +256,49 @@ export class Exchange extends AbstractContract {
         if (_buyOrders.length === 0 || _sellOrders.length === 0) {
             return { buyIds, sellIds, gasEstimate: 0 };
         }
-        const lowestSellPrice: BigNumber = _sellOrders[0].bnPrice;
-        const highestBuyPrice: BigNumber = _buyOrders[0].bnPrice;
+        const lowestSellPrice: BN = _sellOrders[0].price;
+        const highestBuyPrice: BN = _buyOrders[0].price;
 
-        const buyOrders: IBuyOrderCalc[] = _buyOrders
-            .filter((o: IBuyOrder) => o.bnPrice.gte(lowestSellPrice))
-            .map((o: IBuyOrderCalc) => o as IBuyOrderCalc);
+        const buyOrders: IOrder[] = _buyOrders
+            .filter((o: IOrder) => o.price.gte(lowestSellPrice));
 
-        const sellOrders: ISellOrderCalc[] = _sellOrders
-            .filter((o: ISellOrder) => o.bnPrice.lte(highestBuyPrice))
-            .map((o: ISellOrderCalc) => o as ISellOrderCalc);
+        const sellOrders: IOrder[] = _sellOrders
+            .filter((o: IOrder) => o.price.lte(highestBuyPrice));
 
         let buyIdx: number = 0;
         let sellIdx: number = 0;
         let gasEstimate: number = 0;
         let nextGasEstimate: number = MATCH_MULTIPLE_FIRST_MATCH_GAS;
 
+        const E12 = new BN("1000000000000");
```

Comment:
magic number. 
Also does it work only with 2 token DECIMALS ? We avoided so far almost everywhere hardcoding decimals.
 PPM_DIV is also hard coded in this number but that's less likely to change.

Reply:
the point was to replicate the smart contract's code verbatim.

Reply:
i see... so the magic number is there:
```solidity
uint sellWei = sell.amount.mul(uint(price)).mul(E12).roundedDiv(publishedRate);
```
Am I right that means Exchange works only with tokens which have 2 decimals?

Reply:
No. The only assumption is that the `publishedRate` has the same number of decimals as the token.

`E12 = 10^2 = 10^18 / 10^6`, basically 1 ETH in wei (amount) / 1 PPM (price).

Reply:
added a link to the smart contract source.